### PR TITLE
feat: Launch dataset import

### PR DESCRIPTION
**Pull request by MitchellPaff**

### CODE REVIEW

File: datalake/backend/endpoints/dataset_versions/create.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+"""Import dataset-version function."""
```
Comment (imincik) [inline datalake/backend/endpoints/dataset_versions/create.py:1]: This docstring doesn't add too much value to me. I would change it to something more descriptive (like "Dataset versions handler function") or remove it.

File: datalake/backend/endpoints/dataset_versions/create.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+from ..utils import DATASET_TYPES, ENV, JSON_OBJECT, error_response, success_response
+
+stepfunc_client = boto3.client("stepfunctions")
+ssm_client = boto3.client("ssm")
+
+STEP_FUNCTION_ARN_PARAMETER_NAME = f"/{ENV}/StepFuncStateMachineARN"
+
+
+def create_dataset_version(payload: JSON_OBJECT) -> JSON_OBJECT:
+    """POST: Create Dataset."""
```
Comment (imincik) [inline datalake/backend/endpoints/dataset_versions/create.py:21]: This docstring should be changed to "Create Dataset version"

File: datalake/backend/endpoints/dataset_versions/create.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        dataset = DatasetModel.get(
+            hash_key=f"DATASET#{req_body['id']}",
+            range_key=f"TYPE#{req_body['type']}",
+            consistent_read=True,
+        )
+    except DoesNotExist as err:
+        logger.warning(json.dumps({"error": err}, default=str))
+        return error_response(404, f"dataset '{req_body['id']}' could not be found")
+
+    transaction_id = str(uuid.uuid1())
```
Comment (imincik) [inline datalake/backend/endpoints/dataset_versions/create.py:59]: I would rename variable `transaction_id` to `dataset_version_id` or `version_id`

File: datalake/backend/endpoints/dataset_versions/create.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+from jsonschema import ValidationError, validate  # type: ignore[import]
+from pynamodb.exceptions import DoesNotExist
+
+from ..datasets.model import DatasetModel
+from ..utils import DATASET_TYPES, ENV, JSON_OBJECT, error_response, success_response
+
+stepfunc_client = boto3.client("stepfunctions")
+ssm_client = boto3.client("ssm")
+
+STEP_FUNCTION_ARN_PARAMETER_NAME = f"/{ENV}/StepFuncStateMachineARN"
```
Comment (imincik) [inline datalake/backend/endpoints/dataset_versions/create.py:17]: I would rename this variable to something more descriptive like `DATASET_VERSION_CREATION_STEP_FUNCTION ...`

File: datalake/backend/endpoints/datasets/common.py [hunk 5]
```diff
@@ -1,3 +0,0 @@
-from typing import Sequence
```
Comment (imincik) [inline datalake/backend/endpoints/datasets/common.py:1]: I think it would be good to extract this change to separate PR. I like it, but it is out of scope of this PR.

File: datalake/backend/endpoints/datasets/create.py [hunk 6]
```diff
@@ -2,8 +2,7 @@
 
 from jsonschema import ValidationError, validate  # type: ignore[import]
 
-from ..utils import JSON_OBJECT, error_response, success_response
-from .common import DATASET_TYPES
```
Comment (imincik) [inline datalake/backend/endpoints/datasets/create.py:6]: I think it would be good to extract this change to separate PR. I like it, but it is out of scope of this PR.

File: datalake/backend/endpoints/datasets/delete.py [hunk 7]
```diff
@@ -3,8 +3,7 @@
 from jsonschema import ValidationError, validate  # type: ignore[import]
 from pynamodb.exceptions import DoesNotExist
 
-from ..utils import JSON_OBJECT, error_response, success_response
-from .common import DATASET_TYPES
```
Comment (imincik) [inline datalake/backend/endpoints/datasets/delete.py:7]: I think it would be good to extract this change to separate PR. I like it, but it is out of scope of this PR.

File: datalake/backend/endpoints/datasets/get.py [hunk 8]
```diff
@@ -3,8 +3,7 @@
 from jsonschema import ValidationError, validate  # type: ignore[import]
 from pynamodb.exceptions import DoesNotExist
 
-from ..utils import JSON_OBJECT, error_response, success_response
-from .common import DATASET_TYPES
```
Comment (imincik) [inline datalake/backend/endpoints/datasets/get.py:7]: I think it would be good to extract this change to separate PR. I like it, but it is out of scope of this PR.

File: datalake/backend/endpoints/datasets/update.py [hunk 9]
```diff
@@ -3,8 +3,7 @@
 from jsonschema import ValidationError, validate  # type: ignore[import]
 from pynamodb.exceptions import DoesNotExist
 
-from ..utils import JSON_OBJECT, error_response, success_response
```
Comment (imincik) [inline datalake/backend/endpoints/datasets/update.py:6]: I think it would be good to extract this change to separate PR. I like it, but it is out of scope of this PR.

File: datalake/backend/tests/test_check_stac_metadata.py [hunk 10]
```diff
@@ -4,17 +4,16 @@
 from hashlib import sha256, sha512
```
Comment (imincik) [inline datalake/backend/tests/test_check_stac_metadata.py:4]: Looks like most of changes in this file are not related to this PR.

File: datalake/backend/tests/test_lambda_functions.py [hunk 11]
```diff
@@ -1,36 +0,0 @@
-"""
```
Comment (imincik) [inline datalake/backend/tests/test_lambda_functions.py:1]: Are you sure this change should be a part of this PR ?

**Comment by imincik** [comment]
@MitchellPaff , I left some comments to this PR. Would you please rebase this PR against latest master and squash commits which are representing same change ? Please request assign myself for review when done.

Also, I see that you started using Parameter Store. If you really want to keep it as part of this PR I would very much recommend that this PR is quickly followed by PR which is implementing Parameter Store for all other resource names.

Thank you

File: datalake/backend/endpoints/dataset_versions/create.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+            consistent_read=True,
+        )
+    except DoesNotExist as err:
+        logger.warning(json.dumps({"error": err}, default=str))
+        return error_response(404, f"dataset '{req_body['id']}' could not be found")
+
+    transaction_id = str(uuid.uuid1())
+
+    # execute step function
+    stepfunction_input = {
```
Comment (l0b0) [inline datalake/backend/endpoints/dataset_versions/create.py:62]: "step function" is two words, so we should probably name variables accordingly.

File: datalake/backend/endpoints/dataset_versions/create.py (hunk 2 above)
Comment (l0b0) [reply datalake/backend/endpoints/dataset_versions/create.py:21]: Why not remove this docstring? The only thing it tells us in addition to the function name is just that it's responding to a POST request (which the contents of the function doesn't actually care about).

File: datalake/backend/tests/conftest.py [hunk 13]
```diff
@@ -43,6 +44,11 @@ def step_functions_client() -> SFNClient:
     return boto3.client("stepfunctions")
 
 
+@pytest.fixture()
+def ssm_client() -> SSMClient:
```
Comment (l0b0) [inline datalake/backend/tests/conftest.py:48]: This should be before `step_functions_client` alphabetically.

File: datalake/backend/tests/test_check_stac_metadata.py [hunk 14]
```diff
@@ -4,17 +4,16 @@
...
 
 from jsonschema import ValidationError  # type: ignore[import]
-from mypy_boto3_stepfunctions import SFNClient
 from pytest import mark, raises
 from pytest_subtests import SubTests  # type: ignore[import]
 
 from ..endpoints.utils import ResourceName
-from ..processing.check_stac_metadata.task import ProcessingAssetsModel, STACSchemaValidator, main
+from ..processing.check_stac_metadata.task import ProcessingAssetsModel, STACSchemaValidator
+from ..processing.check_stac_metadata.task import main as class_under_test
```
Comment (l0b0) [inline datalake/backend/tests/test_check_stac_metadata.py:16]: `main` is not a class, it's a function. Also, why rename it at import time? There's no other `main` function in this file.

File: datalake/backend/tests/test_dataset_versions_endpoint.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    logger.info("Response: %s", response)
+
+    assert response == {
+        "statusCode": 404,
+        "body": {"message": f"Not Found: dataset '{body['id']}' could not be found"},
+    }
+
+
+@mark.infrastructure
+def test_should_return_success_dataset_exists(
```
Comment (l0b0) [inline datalake/backend/tests/test_dataset_versions_endpoint.py:90]: `test_should_return_success_dataset_exists` should probably be `test_should_return_success_if_dataset_exists`.

File: datalake/backend/tests/test_dataset_versions_endpoint.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+    db_teardown: _pytest.fixtures.FixtureDef[object],  # pylint:disable=unused-argument
+) -> None:
+    # Given a dataset instance
+    dataset_id = "111abc"
```
Comment (l0b0) [inline datalake/backend/tests/test_dataset_versions_endpoint.py:94]: Could this be `any_dataset_id()` instead?

File: datalake/backend/tests/utils.py [hunk 17]
```diff
@@ -120,6 +114,11 @@ def any_dataset_owning_group() -> str:
     return random_string(20)
 
 
+def any_metadata_url() -> str:
+    """Arbitrary-length string"""
+    return random_string(20)
```
Comment (l0b0) [inline datalake/backend/tests/utils.py:119]: That is not a valid URL. Also, would `any_s3_url` work for this case?

**Review by l0b0 — CHANGES_REQUESTED** [review]

File: datalake/backend/endpoints/dataset_versions/create.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+    logger.debug(json.dumps({"response": sfn_response}, default=str))
+
+    # return arn of executing process
+    return success_response(
+        201, {"dataset_version": dataset_version_id, "execution_arn": sfn_response["executionArn"]}
+    )
+
+
+def get_param(parameter: str) -> str:
+    """Retrieve ARN of State Function Machine"""
```
Comment (l0b0) [inline datalake/backend/endpoints/dataset_versions/create.py:83]: Isn't this comment too specific for what the function does?

File: datalake/backend/tests/test_dataset_versions_endpoint.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
+"""
+Dataset Versions endpoint Lambda function tests. Working Data Lake AWS environment is
+required (run '$ cdk deploy' before running tests).
```
Comment (l0b0) [inline datalake/backend/tests/test_dataset_versions_endpoint.py:3]: I think

```
`cdk deploy`
```

would be a more common way to write `'$ cdk deploy'`. The `$` prompt is usually only included with a command when showing a session recording, including the output of commands.

In any case, I think these commands should be `@mark.infrastructure` if they actually require a deployed system. We should probably document that mark in README.md in a separate PR rather than alongside each infrastructure test case.

File: datalake/backend/tests/test_processing_stack.py [hunk 20]
```diff
@@ -45,6 +49,15 @@ class NoComputeEnvironmentFound(Exception):
     pass
 
 
+@mark.infrastructure
+def test_storage_bucket_parameter_created(ssm_client: SSMClient) -> None:
+    """Test if Data Lake State Machine ARN Parameter was created"""
```
Comment (l0b0) [inline datalake/backend/tests/test_processing_stack.py:54]: Should we maybe merge this with the test title to make it clearer? Something like `test_should_create_state_machine_bucket_arn_parameter`?

File: datalake/backend/tests/test_dataset_versions_endpoint.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+from pytest import mark
+
+from ..endpoints.dataset_versions import entrypoint
+from .utils import Dataset, any_dataset_id, any_lambda_context, any_s3_url, any_valid_dataset_type
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def test_should_fail_if_request_not_containing_method() -> None:
```
Comment (l0b0) [inline datalake/backend/tests/test_dataset_versions_endpoint.py:18]: "should fail" is a bad naming convention. Basically, there are an infinite number of ways a piece of code could be said to "fail," and we want the test names to be specific enough to understand which failure mode is under test. In this case, what we are actually verifying is that a missing method results in a "bad request" response, so we could use a name like "test_should_return_required_property_error_when_missing_http_method`.

File: datalake/backend/tests/test_dataset_versions_endpoint_logging.py [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+class LogTests(TestCase):
+    logger: logging.Logger
+
+    @classmethod
+    def setup_class(cls) -> None:
+        cls.logger = logging.getLogger("datalake.backend.endpoints.dataset_versions.create")
+
+    @mark.infrastructure
+    @patch("datalake.backend.endpoints.dataset_versions.create.sfn_client.start_execution")
+    def test_should_log_payload_and_sfn_response(self, start_execution_mock: MagicMock) -> None:
```
Comment (l0b0) [inline datalake/backend/tests/test_dataset_versions_endpoint_logging.py:23]: "sfn" is not necessarily an abbreviation a newbie to the code base would know. I would either expand it or just remove it.

File: datalake/backend/tests/test_dataset_versions_endpoint_logging.py [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+
+            expected_payload_log = dumps({"payload": payload})
+            expected_execution_log = dumps({"response": sfn_response})
+
+            with patch.object(self.logger, "debug") as logger_mock:
+                # when
+                create_dataset_version(payload)
+
+                # then
+                logger_mock.assert_any_call(expected_payload_log)
```
Comment (l0b0) [inline datalake/backend/tests/test_dataset_versions_endpoint_logging.py:45]: If this assertion fails the test doesn't run the second assertion, which can hide another issue. If I understand correctly both assertions require the same setup. If so they should be subtests so that we're sure to run both assertions. If not they should be separate tests.

File: datalake/backend/tests/test_dataset_versions_endpoint_logging.py [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+                logger_mock.assert_any_call(expected_execution_log)
+
+    @patch("datalake.backend.endpoints.dataset_versions.create.validate")
+    def test_should_log_warning_missing_argument(self, validate_schema_mock: MagicMock) -> None:
```
Comment (l0b0) [inline datalake/backend/tests/test_dataset_versions_endpoint_logging.py:49]: Would `test_should_log_missing_argument_warning` be clearer?

File: datalake/backend/tests/test_dataset_versions_endpoint_logging.py [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+
+        with patch.object(self.logger, "warning") as logger_mock:
+            # when
+            create_dataset_version(payload)
+
+            # then
+            logger_mock.assert_any_call(expected_log)
+
+    @patch("datalake.backend.endpoints.dataset_versions.create.DatasetModel.get")
+    def test_should_log_warning_dataset_doesnt_exist(
```
Comment (l0b0) [inline datalake/backend/tests/test_dataset_versions_endpoint_logging.py:71]: Would `test_should_log_warning_if_dataset_does_not_exist` be clearer?

File: datalake/backend/tests/test_lambda_functions.py [hunk 26]
```diff
@@ -34,3 +41,30 @@ def test_should_launch_datasets_endpoint_lambda_function(
...
+    body["metadata-url"] = any_s3_url()
+
+    with Dataset(dataset_id=dataset_id, dataset_type=dataset_type):
+
+        resp = lambda_client.invoke(
+            FunctionName=ResourceName.DATASET_VERSIONS_ENDPOINT_FUNCTION_NAME.value,
+            Payload=json.dumps({"httpMethod": method, "body": body}).encode(),
+            InvocationType="RequestResponse",
+        )
+        json_resp = json.loads(resp["Payload"].read().decode("utf-8"))
```
Comment (l0b0) [inline datalake/backend/tests/test_lambda_functions.py:68]: I *think* this can be simplified to `json_resp = json.load(resp["Payload"])`

File: pyproject.toml [hunk 27]
```diff
@@ -153,6 +153,10 @@ datasets-endpoint = [
     "jsonschema",
     "pynamodb",
 ]
+dataset_versions-endpoint = [
```
Comment (l0b0) [inline pyproject.toml:156]: This extra needs to be mentioned where we run `poetry install` with all the other extras (GitHub Actions, reset-dev-env.bash), even if it overlaps with the others right now.

File: pyproject.toml [hunk 28]
```diff
@@ -122,7 +122,7 @@ shellcheck-py = "*"
 
 [tool.poetry.dev-dependencies.boto3-stubs]
 version = "*"
-extras = ["batch", "dynamodb", "lambda", "s3", "stepfunctions"]
+extras = ["batch", "dynamodb", "lambda", "s3", "stepfunctions", "ssm"]
```
Comment (l0b0) [inline pyproject.toml:125]: `ssm` should be before `stepfunctions`.

**Review by l0b0 — CHANGES_REQUESTED** [review]

File: datalake/backend/tests/test_dataset_versions_endpoint.py [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+    response = entrypoint.lambda_handler({"httpMethod": "POST", "body": body}, any_lambda_context())
+
+    # Then the API should return an error message
+    assert response == {
+        "statusCode": 400,
+        "body": {"message": "Bad Request: 'type' is a required property"},
+    }
+
+
+def test_should_return_error_if_dataset_id_does_not_exist_in_db() -> None:
```
Comment (l0b0) [inline datalake/backend/tests/test_dataset_versions_endpoint.py:67]: This needs `@mark.infrastructure`:

```console
$ pytest -m 'not infrastructure' datalake/
[…]
FAILED datalake/backend/tests/test_dataset_versions_endpoint.py::test_should_return_error_if_dataset_id_does_not_exist_in_db - pynamodb.exceptions.VerboseClientError: An error occurred (Ex...
```

**Review by l0b0 — APPROVED** [review]
:+1:
